rustls/src/crypto/signer.rs

```diff
@@ -61,6 +62,12 @@ pub trait SigningKey: Debug + Send + Sync {
 
     /// What kind of key we have.
     fn algorithm(&self) -> SignatureAlgorithm;
+
+    /// Get the RFC 5280-compliant SubjectPublicKeyInfo (SPKI) of this [`SigningKey`] if it exists.
+    fn public_key(&self) -> Result<Option<Vec<u8>>, Error> {
+        // Opt-out by default
+        Err(Error::UnimplementedSubjectPublicKeyInfo)
```

Comment:
I would say `Ok(None)`?

(I was thinking we'd automatically do this checking, in places where we accept a `CertifiedKey` such as `ResolvesServerCertUsingSni::add` for example. Having the `Option` there codifies the opt-out behaviour, and allows higher levels to avoid raising and then quenching errors.)

Reply:
Makes sense. Changed the default return value to `Ok(None)` in https://github.com/rustls/rustls/pull/1954/commits/38dd19e0ea9471274a9cf67c06b0aad494966a49